File: r0q2/main.js
const readline = require('readline')

function fillSymbolsForWildcard(symbols, wildcard) {
  const length = symbols.length

  const symbolBeforeWildcard = wildcard.startIndex > 0
    ? symbols[wildcard.startIndex - 1]
    : null
  const symbolAfterWildcard = wildcard.startIndex < length - 1
    ? symbols[wildcard.endIndex + 1]
    : null

  if (!symbolBeforeWildcard && !symbolAfterWildcard) {
    return symbols.fill(
      'C',
      wildcard.startIndex,
      wildcard.endIndex + 1
    )
  } else if (!symbolBeforeWildcard) {
    return symbols.fill(
      symbolAfterWildcard,
      wildcard.startIndex,
      wildcard.endIndex + 1
    )
  } else {
    return symbols.fill(
      symbolBeforeWildcard,
      wildcard.startIndex,
      wildcard.endIndex + 1
    )
  }
}

function fillSymbols(symbols) {
  const length = symbols.length

  let modifiedSymbols = symbols.split('')
  let latestWildcard = null

  for (let i = 0; i < length; i++) {
    const symbol = modifiedSymbols[i]
    if (symbol === '?') {
      latestWildcard = {
        startIndex: latestWildcard
          ? latestWildcard.startIndex
          : i,
        endIndex: i
      }
    } else if (latestWildcard) {
      modifiedSymbols = fillSymbolsForWildcard(modifiedSymbols, latestWildcard)
      latestWildcard = null
    }
  }

  if (latestWildcard) {
    modifiedSymbols = fillSymbolsForWildcard(modifiedSymbols, latestWildcard)
    latestWildcard = null
  }

  return modifiedSymbols.join('')
}

function calculateCost({ x, y, symbols }) {
  const length = symbols.length
  let cost = 0

  for (var i = 1; i < symbols.length; i++) {
    const symbolPair = `${symbols[i-1]}${symbols[i]}`
    if (symbolPair === 'CJ') {
      cost += x
    } else if (symbolPair === 'JC') {
      cost += y
    }
  }

  return cost
}

function solveTestCase({ x, y, symbols }) {
  const filledSymbols = fillSymbols(symbols)
  return calculateCost({ x, y, symbols: filledSymbols })
}

function solveProblem(problem) {
  return problem.testCases
    .map(testCase => solveTestCase(testCase))
    .map((result, i) => `Case #${i + 1}: ${result}`)
    .join('\n')
}

function main() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false
  })

  let problem = {
    totalTestCases: 0,
    testCases: [],
  }

  rl.on('line', function (line) {
    if (!line) {
      return
    }
    if (problem.totalTestCases === 0) {
      problem.totalTestCases = Number.parseInt(line)
    } else {
      const tokens = line.split(' ')
      problem.testCases.push({
        x: Number.parseInt(tokens[0]),
        y: Number.parseInt(tokens[1]),
        symbols: tokens[2]
      })
    }
  })
  .on('close', () => {
    console.log(solveProblem(problem))
    process.exit()
  })
}

main()

File: r0q2/main.go
// Moons and Umbrellas
// https://codingcompetitions.withgoogle.com/codejam/round/000000000043580a/00000000006d1145

package main

import (
	"fmt"
	"strings"
)

type TestCase struct {
	CJCost  int
	JCCost  int
	Symbols string
}

type WildcardRange struct {
	StartIndex int
	EndIndex   int
}

func replaceSymbolsBetweenIndexes(symbols string, replacement byte, startIndex int, endIndex int) string {
	return symbols[:startIndex] + strings.Repeat(string(replacement), endIndex-startIndex+1) + symbols[endIndex+1:]
}

func fillBestWildcardsBetweenIndexes(symbols string, startIndex int, endIndex int) string {
	size := len(symbols)

	symbolBeforeWildcard := byte(0)
	symbolAfterWildcard := byte(0)
	if startIndex > 0 {
		symbolBeforeWildcard = symbols[startIndex-1]
	}
	if endIndex < size-1 {
		symbolBeforeWildcard = symbols[endIndex+1]
	}

	if symbolBeforeWildcard == byte(0) && symbolAfterWildcard == byte(0) {
		return replaceSymbolsBetweenIndexes(
			symbols,
			'C',
			startIndex,
			endIndex,
		)
	} else if symbolBeforeWildcard == byte(0) {
		return replaceSymbolsBetweenIndexes(
			symbols,
			symbolAfterWildcard,
			startIndex,
			endIndex,
		)
	} else {
		return replaceSymbolsBetweenIndexes(
			symbols,
			symbolBeforeWildcard,
			startIndex,
			endIndex,
		)
	}
}

func findAllWildcardRanges(symbols string) []*WildcardRange {
	size := len(symbols)
	wildcardRanges := []*WildcardRange{}

	var latestWildcardRange *WildcardRange = nil

	for i := 0; i < size; i++ {
		symbol := symbols[i]
		if symbol == '?' {
			if latestWildcardRange == nil {
				// Found new wildcard range
				latestWildcardRange = &WildcardRange{
					StartIndex: i,
					EndIndex:   i,
				}
				wildcardRanges = append(wildcardRanges, latestWildcardRange)
			} else {
				// Expanding the latest wildcard range
				latestWildcardRange.EndIndex = i
			}
		} else {
			latestWildcardRange = nil
		}
	}

	return wildcardRanges
}

func fillAllWildcards(symbols string) string {
	wildcardRanges := findAllWildcardRanges(symbols)

	filledSymbols := symbols
	for _, wildcardRange := range wildcardRanges {
		filledSymbols = fillBestWildcardsBetweenIndexes(
			filledSymbols,
			wildcardRange.StartIndex,
			wildcardRange.EndIndex,
		)
	}

	return filledSymbols
}

func calculateSymbolsCost(cjCost int, jcCost int, symbols string) int {
	size := len(symbols)
	cost := 0

	for i := 1; i < size; i++ {
		symbolPair := symbols[i-1 : i+1]
		if symbolPair == "CJ" {
			cost += cjCost
		} else if symbolPair == "JC" {
			cost += jcCost
		}
	}

	return cost
}

func solveTestCase(testCase *TestCase) int {
	return calculateSymbolsCost(
		testCase.CJCost,
		testCase.JCCost,
		fillAllWildcards(testCase.Symbols),
	)
}

func readTestCase() *TestCase {
	testCase := &TestCase{}

	fmt.Scanf(
		"%d %d %s",
		&testCase.CJCost,
		&testCase.JCCost,
		&testCase.Symbols,
	)

	return testCase
}

func readTestCases() []*TestCase {
	var totalCases int
	fmt.Scanf("%d", &totalCases)

	testCases := make([]*TestCase, totalCases)
	for i := 0; i < totalCases; i++ {
		testCases[i] = readTestCase()
	}

	return testCases
}

func main() {
	testCases := readTestCases()

	for i, testCase := range testCases {
		answer := solveTestCase(testCase)
		fmt.Printf("Case #%d: %d\n", i+1, answer)
	}
}
